File: shoring/blackjackEngine/src/main/java/com/panopset/blackjackEngine/Deck.java
package com.panopset.blackjackEngine;

import java.util.ArrayList;
import java.util.List;
import com.panopset.compat.Randomop;

public final class Deck {

  private final boolean isBlue;

  private List<Card> shuffled = new ArrayList<>();

  public Deck(boolean isBlue) {
    this.isBlue = isBlue;
    shuffle();
  }

  private void shuffle() {
    List<Card> cards = new ArrayList<>();
    for (CardDefinition cd : CardDefinition.values()) {
      cards.add(new Card(cd, isBlue));
    }
    while (!cards.isEmpty()) {
      shuffled.add(cards.remove(Randomop.random(0, cards.size() - 1)));
    }
  }

  public int remaining() {
    return shuffled.size();
  }

  public Card deal() {
    return shuffled.remove(0);
  }
}

File: code/shoring/blackjackEngine/src/main/kotlin/com/panopset/blackjackEngine/Deck.kt
package com.panopset.blackjackEngine

import com.panopset.compat.Randomop

class Deck(private val isBlue: Boolean) {
    private val shuffled: MutableList<Card> = ArrayList()

    init {
        shuffle()
    }

    private fun shuffle() {
        val cards: MutableList<Card> = ArrayList()
        for (cd in CardDefinition.entries) {
            cards.add(Card(cd, isBlue))
        }
        while (cards.isNotEmpty()) {
            shuffled.add(cards.removeAt(Randomop.random(0, cards.size - 1)))
        }
    }

    fun remaining(): Int {
        return shuffled.size
    }

    fun deal(): Card {
        return shuffled.removeAt(0)
    }
}
